vpr/src/route/route_timing.cpp

```diff
@@ -579,11 +588,17 @@ bool try_timing_driven_route_tmpl(const t_router_opts& router_opts,
             pres_fac = update_pres_fac(router_opts.initial_pres_fac);
         }
 
+        if (budgeting_inf.if_set()) {
+            // If total negative slack is greater than or equal to 0, or RCV is finished, allow router to complete
+            hold_legal = timing_info->hold_total_negative_slack() >= 0 || rcv_finished_count == 0;
+        }
+
         //Have we converged the maximum number of times, did not make any changes, or does it seem
         //unlikely additional convergences will improve QoR?
-        if (legal_convergence_count >= router_opts.max_convergence_count
-            || router_iteration_stats.connections_routed == 0
-            || early_reconvergence_exit_heuristic(router_opts, itry_since_last_convergence, timing_info, best_routing_metrics)) {
+        if ((legal_convergence_count >= router_opts.max_convergence_count
+             || router_iteration_stats.connections_routed == 0
+             || early_reconvergence_exit_heuristic(router_opts, itry_since_last_convergence, timing_info, best_routing_metrics))
```

Comment:
Could move hold_legal into the early_reconvergence_exit_heuristic routine, which I think would be easier to read by simplifying this condition and giving you a place to comment why you're not saying you're exiting early.

Reply:
I agree that would be easier to read, but because the default max_convergence_count is 1, the router would stop early regardless. What I could do is to move all of these conditions into a seperate function?

Reply:
Still outstanding

Reply:
When RCV is selected you could change the legal_convergence_count default to some larger number (we can have dependent option defaults; take a look at the option parsing code and you should find an example).  Then if the early_reconvergence_exit_heuristic understands hold and rcv it can do the right thing.  The if statement won't look as long.

You could also move the whole if into a control routine; that would also work.  But I think changing legal_convergence_count makes sense no matter what, as I believe you've found RCV wants a higher value for this setting.  Better to do that by actually changing the value than writing if statements to work around it.